File: 1006-vowel-spellchecker/vowel-spellchecker.go
func spellchecker(wordlist []string, queries []string) []string {
	// Exact match set
	exact := make(map[string]bool)
	// Lowercase map
	lower := make(map[string]string)
	// Devowel map
	vow := make(map[string]string)

	// Preprocess wordlist
	for _, w := range wordlist {
		exact[w] = true
		lw := lowerIt(w)
		if _, ok := lower[lw]; !ok {
			lower[lw] = w
		}
		dw := devowel(lw)
		if _, ok := vow[dw]; !ok {
			vow[dw] = w
		}
	}

	res := make([]string, len(queries))
	for i, q := range queries {
		if exact[q] {
			// Case 1: Exact match
			res[i] = q
		} else if w, ok := lower[lowerIt(q)]; ok {
			// Case 2: Case-insensitive
			res[i] = w
		} else if w, ok := vow[devowel(lowerIt(q))]; ok {
			// Case 3: Vowel error
			res[i] = w
		} else {
			// Case 4: No match
			res[i] = ""
		}
	}

	return res
}

// Helper function: lowercase
func lowerIt(s string) string {
	return strings.ToLower(s)
}

// Helper function: replace vowels with '#'
func devowel(s string) string {
	vowels := "aeiou"
	result := []rune(s)
    
	for i, ch := range result {
		if strings.ContainsRune(vowels, ch) {
			result[i] = '#'
		}
	}

	return string(result)
}

File: 1006-vowel-spellchecker/vowel-spellchecker.java
class Solution {
    public String[] spellchecker(String[] wordlist, String[] queries) {
        // Exact match set
        Set<String> exact = new HashSet<>();
        // Case-insensitive map
        Map<String, String> lower = new HashMap<>();
        // Vowel-error map
        Map<String, String> vow = new HashMap<>();

        // Preprocess wordlist
        for (String w : wordlist) {
            exact.add(w);
            String lw = lowerIt(w);
            lower.putIfAbsent(lw, w);
            vow.putIfAbsent(devowel(lw), w);
        }

        String[] res = new String[queries.length];
        for (int i = 0; i < queries.length; i++) {
            String q = queries[i];
            if (exact.contains(q))
                // Case 1: Exact match
                res[i] = q;
            else if (lower.containsKey(lowerIt(q)))
                // Case 2: Case-insensitive
                res[i] = lower.get(lowerIt(q));
            else if (vow.containsKey(devowel(lowerIt(q))))
                // Case 3: Vowel error
                res[i] = vow.get(devowel(lowerIt(q)));
            else
                // Case 4: No match
                res[i] = "";
        }

        return res;
    }

    // Helper: lowercase
    private String lowerIt(String s) {
        return s.toLowerCase();
    }

    // Helper: devowel (replace vowels with '#')
    private String devowel(String s) {
        StringBuilder sb = new StringBuilder();
        for (char c : s.toCharArray())
            if ("aeiou".indexOf(c) >= 0)
                sb.append('#');
            else
                sb.append(c);

        return sb.toString();
    }
}